vivisect/__init__.py

```diff
@@ -2617,6 +2617,29 @@ def loadFromFd(self, fd, fmtname=None, baseaddr=None):
 
         return fname
 
+    def loadParsedBin(self, pbin, fmtname=None, baseaddr=None):
```

Comment:
i'm not sure this is ready for prime time.  
please docstring first off.
secondly, what's the purpose of this?  and why are you manually limiting loading to ELF and PE?  i'm really confused.  what's the use-case?

thanks!

Reply:
Sure, a docstring is easy. But aside from that, what part of it disagreeable? It's so that we can load either a top level PE/Elf module into a workspace. Just so we have that link in the chain of loading things and don't want to re-parse a PE/Elf file when we want to load it into a workspace.

And it's limited to ELF and PE because that's what we have top level parser modules for. We don't have a top level module that produces something like an SREC object. If we ever do, we can certainly add that here.

Reply:
i just didn't understand what this was all about.  without the docstr it was difficult to see the value and use-case.  i don't think it being top-level is important, necessarily.  SREC and IHEX could totally be useful here.  MACHO as well, i would imagine.  what happens if we move those into a unified namespace?  the "top-level parser" thing goes away, doesn't it?

Reply:
See, I fail to see how SREC/IHEX/MACHO would even fit in this function, since those don't have top level objects and are already handled by other functions. This is specifically for taking an already parsed PE/Elf object that we've got and loading it into a workspace without having to reparse it.

And sure, if we move to a unified namespace, a lot of things go away, such as this. But this is the first I'm hearing of that at all, so that never factored in. Our API for those have been fairly stable, so I'm not really expecting it to change from what we've got.

Reply:
you had mentioned the desire to have all of Vivisect under one namespace instead of having so many top-levels.  nm, not important.

for macho, a suggestion:
```
    # Instantiate the parser wrapper and parse bytes
    macho = vs_macho.mach_o()
    macho.vsParse(filebytes)
```
the pbin becomes that `macho` object.
that may take a small amount of wrapper to keep it in line with Elf/PE, or it may not.

for SREC:
```
    srec = v_srec.SRecFile()
    with open(filename, 'rb') as f:
        shdr = f.read(offset)
        sbytes = f.read()
        if offset:
            logger.debug('skipping %d bytes: %r', offset, repr(shdr)[:300])

        fname = vw.addFile(filename, 0, v_parsers.md5Bytes(shdr + sbytes))
        vw.setFileMeta(fname, 'sha256', v_parsers.sha256Bytes(shdr + sbytes))

        srec.vsParse(sbytes)
```
that srec object is the pbin.

IHEX is virtually the same.

the point, i thought, was to keep us from having to reparse the binary through the parser object manually?

i get that this is probably more than what you wanted to do for this PR... we can talk about it being "in a future PR" but leaving them out of this goodness seems shortsighted and kludgy.  yes another thing where Viv is inconsistent and incomplete.

Reply:
Oh wow yea. I did. That was a long time ago though. I had forgotten that I wanted that at one point.

And I mostly kept away from those since they would require deep changes to all those parsers , which is  more than what I wanted to do for this PR.

But I do want this convenience function in here since it provides a nice bridge from parsing to workspace that I still feel like we don't have.

And yea, I know that code is in the parsers for those. But I wanted to avoid big changes to those parsers since a lot of them are "lightly" tested right now, and I don't have the time to make tests for them